### PR TITLE
fix(config): extend validation to all properties

### PR DESCRIPTION
**Pull request by Phil91**
## Description

validation for custom attributes EnumEnumerable and DistinctValues extended to validate Properties of properties as well added base implementation for custom enumerable validators

## Why

The current validatio implmentation only validates properties within the given class, but doesn't validate properties of properties.

## Issue

N/A - Jira Issue: CPLP-2698

## Checklist

Please delete options that are not relevant.

- [x] I have followed the [contributing guidelines](https://github.com/eclipse-tractusx/portal-assets/blob/main/developer/Technical%20Documentation/Dev%20Process/How%20to%20contribute.md#commit-and-pr-guidelines)
- [x] I have performed a self-review of my own code
- [x] I have successfully tested my changes locally
- [x] I have added tests that prove my changes work
- [x] I have checked that new and existing tests pass locally with my changes
- [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/framework/Framework.Models/Validation/BaseOptionEnumerableValidation.cs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        {
+            yield return validationResult;
+        }
+    }
+
+    protected abstract IEnumerable<ValidationResult> ValidateAttribute(IConfiguration config, PropertyInfo property, string propertyName);
+
+    private IEnumerable<ValidationResult> CheckPropertiesOfProperty(IConfiguration configSection, PropertyInfo property, string propertyName)
+    {
+        var validationResults = new List<ValidationResult>();
```
Comment (ntruchsess) [inline src/framework/Framework.Models/Validation/BaseOptionEnumerableValidation.cs:105]: same here:
```
        if (property.PropertyType.IsClass)
        {
            foreach (var result in GetValidationErrors(property.PropertyType, configSection.GetSection(propertyName)))
            {
                yield return result;
            }
        }
        else if (property.PropertyType.GetInterfaces().Contains(typeof(IEnumerable)))
        {
            var propertyType = property.PropertyType.GetGenericArguments().FirstOrDefault();
            if (propertyType == null || IgnoreTypes.Contains(propertyType) || propertyType.IsEnum)
                yield break;

            // Validate each item of the enumerable
            var listValues = (IEnumerable)configSection.GetSection(propertyName).Get(property.PropertyType);
            if (listValues != null)
            {
                foreach (var result in Enumerable.Range(0, listValues.ToIEnumerable().Count())
                    .SelectMany(i => GetValidationErrors(propertyType, configSection.GetSection($"{propertyName}:{i}"))))
                {
                    yield return result;
                }
            }
        }
```
or using pure LINQ:
```
    private IEnumerable<ValidationResult> CheckPropertiesOfProperty(IConfiguration configSection, PropertyInfo property, string propertyName) =>
        property.PropertyType switch
        {
            var x when x.IsClass => GetValidationErrors(property.PropertyType, configSection.GetSection(propertyName)),
            var x when x.GetInterfaces().Contains(typeof(IEnumerable)) &&
                    x.GetGenericArguments()
                        .Where(genericType => !genericType.IsEnum)
                        .Except(IgnoreTypes)
                        .IfAny(genericTypes => genericTypes.First(), out var genericType) =>
                            (configSection.GetSection(propertyName).Get(property.PropertyType) as IEnumerable)
                                ?.ToIEnumerable()
                                .Select((_, i) => configSection.GetSection($"{propertyName}:{i}"))
                                .SelectMany(section => GetValidationErrors(genericType!, section)),
            _ => null
        } ?? Enumerable.Empty<ValidationResult>();
```

File: src/framework/Framework.Web/WebApplicationBuildRunner.cs [hunk 2]
```diff
@@ -47,10 +47,12 @@ public static class WebApplicationBuildRunner
         }
         catch (Exception ex) when (!ex.GetType().Name.Equals("StopTheHostException", StringComparison.Ordinal))
         {
+            LoggingExtensions.EnsureInitialized();
```
Comment (ntruchsess) [inline src/framework/Framework.Web/WebApplicationBuildRunner.cs:50]: I guess I'm missing something here. Please shed some light on why the LoggingExtensions need to be initialized a second time.

File: tests/framework/Framework.Models.Tests/ValidationTests.cs [hunk 3]
```diff
@@ -61,19 +61,33 @@ public class InvalidDistinctValuesTestSettings
     public void DistinctValuesValidation_Distinct_ReturnsExpected()
     {
         // Arrange
-        var settings = new DistinctValuesTestSettings()
+        const string configuration = @"
```
Comment (ntruchsess) [inline tests/framework/Framework.Models.Tests/ValidationTests.cs:64]: some tests that validates the recursive validation of nested properties would be great. So far those tests were also passed by the previous validator implementation

**Review by ntruchsess — CHANGES_REQUESTED** [review]

File: src/framework/Framework.Models/Validation/BaseOptionEnumerableValidation.cs [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        {
+            var propertyType = property.PropertyType.GetGenericArguments().FirstOrDefault();
+            if (propertyType == null || IgnoreTypes.Contains(propertyType) || propertyType.IsEnum)
+                yield break;
+
+            // Validate each item of the enumerable
+            var listValues = (IEnumerable)configSection.GetSection(propertyName).Get(property.PropertyType);
+            if (listValues != null)
+            {
+                var errors = Enumerable.Range(0, listValues.ToIEnumerable().Count())
```
Comment (ntruchsess) [inline src/framework/Framework.Models/Validation/BaseOptionEnumerableValidation.cs:120]: This 'Enumerable.Range(0...Count()) is a cumbersome way to generate index values that are implicitly provided by Select and SelectMany (both are overloaded, see my previous remark where I use it in the pure linq implementation)

**Review by ntruchsess — CHANGES_REQUESTED** [review]

File: tests/framework/Framework.Models.Tests/ValidationTests.cs (hunk 3 above)
Comment (Phil91) [reply tests/framework/Framework.Models.Tests/ValidationTests.cs:69]: added some test cases for invalid properties within a property

File: src/framework/Framework.Models/Validation/BaseOptionEnumerableValidation.cs [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+using System.Collections;
+using System.Collections.Immutable;
+using System.ComponentModel.DataAnnotations;
+using System.Reflection;
+
+namespace Org.Eclipse.TractusX.Portal.Backend.Framework.Models.Validation;
+
+public abstract class BaseOptionEnumerableValidation<TOptions> : IValidateOptions<TOptions> where TOptions : class
+{
+    protected ImmutableArray<Type> IgnoreTypes = new List<Type>
```
[the remark this answers is not in the document]
Comment (Phil91) [reply src/framework/Framework.Models/Validation/BaseOptionEnumerableValidation.cs]: changed.

File: src/framework/Framework.Models/Validation/BaseOptionEnumerableValidation.cs [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        var validationErrors = GetValidationErrors(options.GetType(), Section);
+
+        return validationErrors.IfAny(
+            errors => errors.Select(r => $"DataAnnotation validation failed for members: '{string.Join(",", r.MemberNames)}' with the error: '{r.ErrorMessage}'."),
+            out var messages)
+                ? ValidateOptionsResult.Fail(messages)
+                : ValidateOptionsResult.Success;
+    }
+
+    private IEnumerable<ValidationResult> GetValidationErrors(Type type, IConfiguration configSection)
```
[the remark this answers is not in the document]
Comment (Phil91) [reply src/framework/Framework.Models/Validation/BaseOptionEnumerableValidation.cs]: really like the LINQ implementation!

File: src/framework/Framework.Models/Validation/BaseOptionEnumerableValidation.cs (hunk 1 above)
Comment (Phil91) [reply src/framework/Framework.Models/Validation/BaseOptionEnumerableValidation.cs]: changed as well.

File: src/framework/Framework.Models/Validation/BaseOptionEnumerableValidation.cs (hunk 4 above)
Comment (Phil91) [reply src/framework/Framework.Models/Validation/BaseOptionEnumerableValidation.cs]: somehow i missed the .select part to get the index

File: src/framework/Framework.Web/WebApplicationBuildRunner.cs (hunk 2 above)
Comment (Phil91) [reply src/framework/Framework.Web/WebApplicationBuildRunner.cs:50]: need to look deeper into that as well. just saw that if we don't have this line, the logger won't write any logs when an error occurs and we'll hit the catch

**Review by ntruchsess — DISMISSED** [review]

**Review by ntruchsess — APPROVED** [review]

**Pull request merged**